1 - 注册表单的实现 - 20201127/lvyan_main.dart

```diff
@@ -0,0 +1,196 @@
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+
+void main() {
+  runApp(MyApp());
+}
+
+class MyApp extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: 'Flutter Demo',
+      theme: ThemeData(
+        primarySwatch: Colors.blue,
+        visualDensity: VisualDensity.adaptivePlatformDensity,
+      ),
+      home: const MyHomePage(title: 'Flutter Demo Home Page'),
+    );
+  }
+}
+
+class MyHomePage extends StatefulWidget {
+  const MyHomePage({Key key, this.title}) : super(key: key);
+
+  final String title;
+
+  @override
+  _MyHomePageState createState() => _MyHomePageState();
+}
+
+class _MyHomePageState extends State<MyHomePage> {
+  final TextEditingController _usernameController = TextEditingController();
+  final TextEditingController _passwordController = TextEditingController();
+
+  bool _visible = false;
+
+  bool _isLogin = false;
+
+  final StreamController<String> _usernameStream =
+      StreamController<String>.broadcast();
+  final StreamController<String> _passwordStream =
+      StreamController<String>.broadcast();
+
+  /// 不知道怎么合并两个流的结果 😄
+  final StreamController<bool> _loginStream = StreamController<bool>();
+
+  Timer _loginTimer;
+
+  @override
+  void initState() {
+    _usernameController.addListener(() {
+      _usernameStream.sink.add(_usernameController.text);
+
+      _loginStream.sink.add(_usernameController.text.isNotEmpty &&
+          _passwordController.text.length >= 8 &&
+          _passwordController.text.split('').toSet().length >= 6);
+    });
+    _passwordController.addListener(() {
+      _passwordStream.sink.add(_passwordController.text);
+
+      _loginStream.sink.add(_usernameController.text.isNotEmpty &&
+          _passwordController.text.length >= 8 &&
+          _passwordController.text.split('').toSet().length >= 6);
+    });
+    super.initState();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final Widget clearIcon = StreamBuilder<String>(
```

Comment:
What's the difference between `Widget get clearIcon` placed outside and `final Widget clearIcon` placed inside in this case?